File: src/domain/models/fitness_test_suite.py
"""FitnessTestSuite aggregate root for the Architecture Testing bounded context.

Generates import-linter contracts and pytestarch rules from bounded contexts
and their subdomain classifications. Enforces invariants per DDD.md Section 5.

Invariants:
1. Every bounded context must have at least one contract.
2. Core subdomains must have all 4 contract types.
3. Supporting subdomains must have layers + forbidden.
4. Generic subdomains must have at least forbidden.
5. No contract may reference modules outside its bounded context.
"""

from __future__ import annotations

import uuid
from typing import TYPE_CHECKING

from src.domain.models.errors import InvariantViolationError
from src.domain.models.fitness_values import (
    ArchRule,
    Contract,
    ContractStrictness,
    ContractType,
)

if TYPE_CHECKING:
    from src.domain.events.fitness_events import FitnessTestsGenerated
    from src.domain.models.domain_values import BoundedContext


def _snake_case(name: str) -> str:
    """Convert a PascalCase, space-separated, or hyphenated name to snake_case.

    Handles consecutive uppercase gracefully: ``ABCTest`` → ``abc_test``.
    """
    result: list[str] = []
    for i, ch in enumerate(name):
        if ch.isupper():
            if result and result[-1] != "_":
                # Only insert underscore before an uppercase char when:
                # - previous char was lowercase, OR
                # - next char is lowercase (end of acronym like ABCTest → abc_test)
                prev_lower = name[i - 1].islower() if i > 0 else False
                next_lower = name[i + 1].islower() if i + 1 < len(name) else False
                if prev_lower or (next_lower and i > 0 and name[i - 1].isupper()):
                    result.append("_")
            result.append(ch.lower())
        elif ch in (" ", "-"):
            if result and result[-1] != "_":
                result.append("_")
        else:
            result.append(ch)
    return "".join(result)


class FitnessTestSuite:
    """Aggregate root: generates and manages architecture fitness tests.

    Attributes:
        suite_id: Unique identifier for this suite.
        root_package: The root Python package being tested.
    """

    def __init__(self, root_package: str) -> None:
        self.suite_id: str = str(uuid.uuid4())
        self.root_package: str = root_package
        self._contracts: list[Contract] = []
        self._arch_rules: list[ArchRule] = []
        self._events: list[FitnessTestsGenerated] = []
        self._approved: bool = False

    # -- Properties -----------------------------------------------------------

    @property
    def contracts(self) -> tuple[Contract, ...]:
        """All generated contracts (defensive copy)."""
        return tuple(self._contracts)

    @property
    def arch_rules(self) -> tuple[ArchRule, ...]:
        """All generated pytestarch rules (defensive copy)."""
        return tuple(self._arch_rules)

    @property
    def events(self) -> tuple[FitnessTestsGenerated, ...]:
        """Domain events produced by this aggregate (defensive copy)."""
        return tuple(self._events)

    # -- Commands -------------------------------------------------------------

    def generate_contracts(
        self,
        bounded_contexts: tuple[BoundedContext, ...],
    ) -> None:
        """Generate contracts for each bounded context based on classification.

        Args:
            bounded_contexts: The classified bounded contexts.

        Raises:
            InvariantViolationError: If suite is already approved, no bounded
                contexts provided, or any BC lacks classification.
        """
        if self._approved:
            msg = "Cannot regenerate contracts on an approved suite"
            raise InvariantViolationError(msg)

        if not bounded_contexts:
            msg = "No bounded contexts to generate fitness tests for"
            raise InvariantViolationError(msg)

        self._contracts.clear()
        self._arch_rules.clear()

        for bc in bounded_contexts:
            if bc.classification is None:
                msg = f"Bounded context '{bc.name}' has no subdomain classification"
                raise InvariantViolationError(msg)

            strictness = ContractStrictness.from_classification(bc.classification)
            required_types = strictness.required_contract_types()
            module_prefix = f"{self.root_package}.{_snake_case(bc.name)}"

            for ct in required_types:
                contract = self._build_contract(ct, bc.name, module_prefix)
                self._contracts.append(contract)

            # Generate pytestarch rule for each BC
            self._arch_rules.append(
                ArchRule(
                    name=f"{bc.name} domain isolation",
                    assertion=(
                        f"modules in {module_prefix}.domain should not "
                        f"import from {module_prefix}.infrastructure"
                    ),
                    context_name=bc.name,
                )
            )

    def preview(self) -> str:
        """Return a human-readable preview of generated contracts.

        Raises:
            InvariantViolationError: If no contracts have been generated.
        """
        if not self._contracts:
            msg = "No contracts generated yet — call generate_contracts() first"
            raise InvariantViolationError(msg)

        lines: list[str] = [
            f"Fitness Test Suite: {self.root_package}",
            f"Total contracts: {len(self._contracts)}",
            f"Total arch rules: {len(self._arch_rules)}",
            "",
        ]

        # Group by context
        contexts_seen: dict[str, list[Contract]] = {}
        for c in self._contracts:
            contexts_seen.setdefault(c.context_name, []).append(c)

        for ctx_name, contracts in contexts_seen.items():
            strictness = self._infer_strictness(contracts)
            types = [c.contract_type.value for c in contracts]
            lines.append(f"  {ctx_name} ({strictness.value.upper()}):")
            lines.append(f"    Contracts: {', '.join(types)}")
            lines.append("")

        return "\n".join(lines)

    def approve(self) -> None:
        """Approve the suite, emitting FitnessTestsGenerated.

        Raises:
            InvariantViolationError: If suite has no contracts, is already
                approved, or any contract references modules outside its BC.
        """
        if self._approved:
            msg = "Suite already approved"
            raise InvariantViolationError(msg)

        if not self._contracts:
            msg = "Cannot approve suite with no contracts"
            raise InvariantViolationError(msg)

        self._validate_module_boundaries()

        self._approved = True

        from src.domain.events.fitness_events import FitnessTestsGenerated

        self._events.append(
            FitnessTestsGenerated(
                suite_id=self.suite_id,
                root_package=self.root_package,
                contracts=tuple(self._contracts),
                arch_rules=tuple(self._arch_rules),
            )
        )

    # -- Rendering ------------------------------------------------------------

    def render_import_linter_toml(self) -> str:
        """Render import-linter TOML configuration.

        Returns:
            Valid TOML string for [tool.importlinter] section.

        Raises:
            InvariantViolationError: If no contracts generated.
        """
        if not self._contracts:
            msg = "No contracts generated yet — call generate_contracts() first"
            raise InvariantViolationError(msg)

        lines: list[str] = [
            "[tool.importlinter]",
            f'root_package = "{self.root_package}"',
            "",
        ]

        for c in self._contracts:
            lines.append("[[tool.importlinter.contracts]]")
            lines.append(f'name = "{c.name}"')
            lines.append(f'type = "{c.contract_type.value}"')

            if c.contract_type == ContractType.LAYERS:
                lines.append("layers = [")
                lines.extend(f'  "{m}",' for m in c.modules)
                lines.append("]")
            elif c.contract_type == ContractType.FORBIDDEN:
                lines.append("source_modules = [")
                lines.extend(f'  "{m}",' for m in c.modules)
                lines.append("]")
                if c.forbidden_modules:
                    lines.append("forbidden_modules = [")
                    lines.extend(f'  "{m}",' for m in c.forbidden_modules)
                    lines.append("]")
            elif c.contract_type == ContractType.INDEPENDENCE:
                lines.append("modules = [")
                lines.extend(f'  "{m}",' for m in c.modules)
                lines.append("]")
            elif c.contract_type == ContractType.ACYCLIC_SIBLINGS:
                lines.append(f'source_module = "{c.modules[0]}"')

            lines.append("")

        return "\n".join(lines)

    def render_pytestarch_tests(self) -> str:
        """Render pytestarch test file content.

        Returns:
            Valid Python test file using pytestarch.

        Raises:
            InvariantViolationError: If no contracts generated.
        """
        if not self._contracts:
            msg = "No contracts generated yet — call generate_contracts() first"
            raise InvariantViolationError(msg)

        lines: list[str] = [
            '"""Auto-generated architecture fitness tests.',
            "",
            "Generated by alty from bounded context map.",
            '"""',
            "",
            "import pytest",
            "from pytestarch import get_evaluable_architecture, Rule",
            "",
            "",
            '@pytest.fixture(scope="session")',
            "def evaluable():",
            f'    return get_evaluable_architecture(".", "{self.root_package}")',
            "",
        ]

        for rule in self._arch_rules:
            fn_name = f"test_{_snake_case(rule.name)}"
            lines.append("")
            lines.append(f"def {fn_name}(evaluable):")
            lines.append(f'    """{rule.assertion}"""')

            # Extract module names from assertion
            ctx_module = _snake_case(rule.context_name)
            domain_mod = f"{self.root_package}.{ctx_module}.domain"
            infra_mod = f"{self.root_package}.{ctx_module}.infrastructure"

            lines.append("    rule = (")
            lines.append("        Rule()")
            lines.append("        .modules_that()")
            lines.append(f'        .are_named("{infra_mod}")')
            lines.append("        .should_not()")
            lines.append("        .be_imported_by_modules_that()")
            lines.append(f'        .are_named("{domain_mod}")')
            lines.append("    )")
            lines.append("    rule.assert_applies(evaluable)")
            lines.append("")

        return "\n".join(lines)

    # -- Private helpers ------------------------------------------------------

    def _validate_module_boundaries(self) -> None:
        """Invariant 5: no contract may reference modules outside its BC.

        Raises:
            InvariantViolationError: If any contract references a module
                outside the expected prefix for its bounded context.
        """
        for contract in self._contracts:
            expected_prefix = f"{self.root_package}.{_snake_case(contract.context_name)}"
            for module in contract.modules:
                if not module.startswith(expected_prefix):
                    msg = (
                        f"Contract '{contract.name}' references module "
                        f"'{module}' outside its bounded context "
                        f"'{contract.context_name}' (expected prefix: "
                        f"'{expected_prefix}')"
                    )
                    raise InvariantViolationError(msg)
            for module in contract.forbidden_modules:
                if not module.startswith(expected_prefix):
                    msg = (
                        f"Contract '{contract.name}' references forbidden "
                        f"module '{module}' outside its bounded context "
                        f"'{contract.context_name}' (expected prefix: "
                        f"'{expected_prefix}')"
                    )
                    raise InvariantViolationError(msg)

    def _build_contract(
        self,
        contract_type: ContractType,
        context_name: str,
        module_prefix: str,
    ) -> Contract:
        """Build a single contract for a bounded context."""
        if contract_type == ContractType.LAYERS:
            return Contract(
                name=f"{context_name} DDD layer contract",
                contract_type=ContractType.LAYERS,
                context_name=context_name,
                modules=(
                    f"{module_prefix}.infrastructure",
                    f"{module_prefix}.application",
                    f"{module_prefix}.domain",
                ),
            )

        if contract_type == ContractType.FORBIDDEN:
            return Contract(
                name=f"{context_name} domain isolation",
                contract_type=ContractType.FORBIDDEN,
                context_name=context_name,
                modules=(f"{module_prefix}.domain",),
                forbidden_modules=(f"{module_prefix}.infrastructure",),
            )

        if contract_type == ContractType.INDEPENDENCE:
            return Contract(
                name=f"{context_name} independence",
                contract_type=ContractType.INDEPENDENCE,
                context_name=context_name,
                modules=(module_prefix,),
            )

        # ACYCLIC_SIBLINGS
        return Contract(
            name=f"{context_name} acyclic siblings",
            contract_type=ContractType.ACYCLIC_SIBLINGS,
            context_name=context_name,
            modules=(module_prefix,),
        )

    @staticmethod
    def _infer_strictness(contracts: list[Contract]) -> ContractStrictness:
        """Infer strictness level from contract types present."""
        types = {c.contract_type for c in contracts}
        if len(types) >= 4:
            return ContractStrictness.STRICT
        if ContractType.LAYERS in types:
            return ContractStrictness.MODERATE
        return ContractStrictness.MINIMAL

File: internal/fitness/domain/fitness_test_suite.go
package domain

import (
	"fmt"
	"strings"

	domainerrors "github.com/alto-cli/alto/internal/shared/domain/errors"
	"github.com/alto-cli/alto/internal/shared/domain/identity"
	vo "github.com/alto-cli/alto/internal/shared/domain/valueobjects"
)

// BoundedContextInput is the input data for generating fitness tests.
type BoundedContextInput struct {
	Classification *vo.SubdomainClassification
	Name           string
	Responsibility string
}

// SnakeCase converts a PascalCase, space-separated, or hyphenated name to snake_case.
func SnakeCase(name string) string {
	var result []byte
	for i := 0; i < len(name); i++ {
		ch := name[i]
		switch {
		case ch >= 'A' && ch <= 'Z':
			if len(result) > 0 && result[len(result)-1] != '_' {
				prevLower := i > 0 && name[i-1] >= 'a' && name[i-1] <= 'z'
				nextLower := i+1 < len(name) && name[i+1] >= 'a' && name[i+1] <= 'z'
				if prevLower || (nextLower && i > 0 && name[i-1] >= 'A' && name[i-1] <= 'Z') {
					result = append(result, '_')
				}
			}
			result = append(result, ch+32) // toLower
		case ch == ' ' || ch == '-':
			if len(result) > 0 && result[len(result)-1] != '_' {
				result = append(result, '_')
			}
		default:
			result = append(result, ch)
		}
	}
	return string(result)
}

// FitnessTestSuite is the aggregate root for architecture fitness tests.
type FitnessTestSuite struct {
	suiteID     string
	rootPackage string
	contracts   []Contract
	archRules   []ArchRule
	events      []FitnessTestsGenerated
	approved    bool
}

// NewFitnessTestSuite creates a new FitnessTestSuite aggregate root.
func NewFitnessTestSuite(rootPackage string) *FitnessTestSuite {
	return &FitnessTestSuite{
		suiteID:     identity.NewID(),
		rootPackage: rootPackage,
	}
}

// SuiteID returns the unique suite identifier.
func (s *FitnessTestSuite) SuiteID() string { return s.suiteID }

// RootPackage returns the root package name.
func (s *FitnessTestSuite) RootPackage() string { return s.rootPackage }

// Contracts returns a defensive copy.
func (s *FitnessTestSuite) Contracts() []Contract {
	out := make([]Contract, len(s.contracts))
	copy(out, s.contracts)
	return out
}

// ArchRules returns a defensive copy.
func (s *FitnessTestSuite) ArchRules() []ArchRule {
	out := make([]ArchRule, len(s.archRules))
	copy(out, s.archRules)
	return out
}

// Events returns a defensive copy.
func (s *FitnessTestSuite) Events() []FitnessTestsGenerated {
	out := make([]FitnessTestsGenerated, len(s.events))
	copy(out, s.events)
	return out
}

// GenerateContracts generates contracts for each bounded context based on classification.
func (s *FitnessTestSuite) GenerateContracts(bcs []BoundedContextInput) error {
	if s.approved {
		return fmt.Errorf("cannot regenerate contracts on an approved suite: %w",
			domainerrors.ErrInvariantViolation)
	}
	if len(bcs) == 0 {
		return fmt.Errorf("no bounded contexts to generate fitness tests for: %w",
			domainerrors.ErrInvariantViolation)
	}

	s.contracts = nil
	s.archRules = nil

	for _, bc := range bcs {
		if bc.Classification == nil {
			return fmt.Errorf("bounded context '%s' has no subdomain classification: %w",
				bc.Name, domainerrors.ErrInvariantViolation)
		}

		strictness := StrictnessFromClassification(*bc.Classification)
		requiredTypes := RequiredContractTypes(strictness)
		modulePrefix := s.rootPackage + "." + SnakeCase(bc.Name)

		for _, ct := range requiredTypes {
			contract := buildContract(ct, bc.Name, modulePrefix)
			s.contracts = append(s.contracts, contract)
		}

		s.archRules = append(s.archRules, NewArchRule(
			bc.Name+" domain isolation",
			fmt.Sprintf("modules in %s.domain should not import from %s.infrastructure",
				modulePrefix, modulePrefix),
			bc.Name,
		))
	}
	return nil
}

// Preview returns a human-readable preview of generated contracts.
func (s *FitnessTestSuite) Preview() (string, error) {
	if len(s.contracts) == 0 {
		return "", fmt.Errorf("no contracts generated yet — call GenerateContracts() first: %w",
			domainerrors.ErrInvariantViolation)
	}

	var b strings.Builder
	fmt.Fprintf(&b, "Fitness Test Suite: %s\n", s.rootPackage)
	fmt.Fprintf(&b, "Total contracts: %d\n", len(s.contracts))
	fmt.Fprintf(&b, "Total arch rules: %d\n\n", len(s.archRules))

	contextsSeen := make(map[string][]Contract)
	for _, c := range s.contracts {
		contextsSeen[c.contextName] = append(contextsSeen[c.contextName], c)
	}

	for ctxName, contracts := range contextsSeen {
		strictness := inferStrictness(contracts)
		types := make([]string, len(contracts))
		for i, c := range contracts {
			types[i] = string(c.contractType)
		}
		fmt.Fprintf(&b, "  %s (%s):\n", ctxName, strings.ToUpper(string(strictness)))
		fmt.Fprintf(&b, "    Contracts: %s\n\n", strings.Join(types, ", "))
	}

	return b.String(), nil
}

// Approve approves the suite, emitting FitnessTestsGenerated.
func (s *FitnessTestSuite) Approve() error {
	if s.approved {
		return fmt.Errorf("suite already approved: %w", domainerrors.ErrInvariantViolation)
	}
	if len(s.contracts) == 0 {
		return fmt.Errorf("cannot approve suite with no contracts: %w",
			domainerrors.ErrInvariantViolation)
	}

	if err := s.validateModuleBoundaries(); err != nil {
		return err
	}

	s.approved = true
	s.events = append(s.events, NewFitnessTestsGenerated(
		s.suiteID, s.rootPackage,
		s.contracts, s.archRules,
	))
	return nil
}

// RenderImportLinterTOML renders import-linter TOML configuration.
func (s *FitnessTestSuite) RenderImportLinterTOML() (string, error) {
	if len(s.contracts) == 0 {
		return "", fmt.Errorf("no contracts generated yet — call GenerateContracts() first: %w",
			domainerrors.ErrInvariantViolation)
	}

	var b strings.Builder
	b.WriteString("[tool.importlinter]\n")
	fmt.Fprintf(&b, "root_package = \"%s\"\n\n", s.rootPackage)

	for _, c := range s.contracts {
		b.WriteString("[[tool.importlinter.contracts]]\n")
		fmt.Fprintf(&b, "name = \"%s\"\n", c.name)
		fmt.Fprintf(&b, "type = \"%s\"\n", string(c.contractType))

		switch c.contractType {
		case ContractTypeLayers:
			b.WriteString("layers = [\n")
			for _, m := range c.modules {
				fmt.Fprintf(&b, "  \"%s\",\n", m)
			}
			b.WriteString("]\n")
		case ContractTypeForbidden:
			b.WriteString("source_modules = [\n")
			for _, m := range c.modules {
				fmt.Fprintf(&b, "  \"%s\",\n", m)
			}
			b.WriteString("]\n")
			if len(c.forbiddenModules) > 0 {
				b.WriteString("forbidden_modules = [\n")
				for _, m := range c.forbiddenModules {
					fmt.Fprintf(&b, "  \"%s\",\n", m)
				}
				b.WriteString("]\n")
			}
		case ContractTypeIndependence:
			b.WriteString("modules = [\n")
			for _, m := range c.modules {
				fmt.Fprintf(&b, "  \"%s\",\n", m)
			}
			b.WriteString("]\n")
		case ContractTypeAcyclicSiblings:
			if len(c.modules) > 0 {
				fmt.Fprintf(&b, "source_module = \"%s\"\n", c.modules[0])
			}
		}
		b.WriteString("\n")
	}

	return b.String(), nil
}

// RenderArchGoYAML renders arch-go.yml configuration from a BoundedContextMap.
// threshold specifies the compliance percentage (100 for greenfield, 80 for brownfield).
func (s *FitnessTestSuite) RenderArchGoYAML(bcMap *BoundedContextMap, threshold int) (string, error) {
	if bcMap == nil || len(bcMap.Contexts()) == 0 {
		return "", fmt.Errorf("bounded context map is empty: %w", domainerrors.ErrInvariantViolation)
	}

	rootPkg := bcMap.RootPackage()
	var b strings.Builder

	// Header
	b.WriteString("version: 1\n\n")
	b.WriteString("threshold:\n")
	fmt.Fprintf(&b, "  compliance: %d\n", threshold)
	fmt.Fprintf(&b, "  coverage: %d\n\n", threshold)

	b.WriteString("dependenciesRules:\n")

	// Build context name set for cross-context isolation
	contextModules := make(map[string]string) // name -> modulePath
	for _, ctx := range bcMap.Contexts() {
		contextModules[ctx.Name()] = ctx.ModulePath()
	}

	// Build allowed dependencies from relationships
	allowedDeps := make(map[string]map[string]bool) // source -> set of allowed targets
	for _, ctx := range bcMap.Contexts() {
		allowedDeps[ctx.Name()] = make(map[string]bool)
		for _, rel := range ctx.Relationships() {
			if rel.Direction() == RelationshipUpstream {
				// Upstream means this context depends on target
				allowedDeps[ctx.Name()][rel.Target()] = true
			}
		}
	}

	// Helper to build full package path
	fullPkg := func(modulePath, layer string) string {
		if layer == "" {
			return fmt.Sprintf("%s/internal/%s", rootPkg, modulePath)
		}
		return fmt.Sprintf("%s/internal/%s/%s", rootPkg, modulePath, layer)
	}

	// Generate layer rules for each context
	for _, ctx := range bcMap.Contexts() {
		modulePath := ctx.ModulePath()

		// Domain layer: shouldOnlyDependsOn + shouldNotDependsOn
		b.WriteString("  # " + ctx.Name() + " domain layer isolation\n")
		fmt.Fprintf(&b, "  - package: %q\n", fullPkg(modulePath, "domain"))
		b.WriteString("    shouldOnlyDependsOn:\n")
		b.WriteString("      internal:\n")
		fmt.Fprintf(&b, "        - %q\n", fullPkg(modulePath, "domain"))
		fmt.Fprintf(&b, "        - %q\n", fullPkg("shared/domain", ""))
		b.WriteString("      external:\n")
		b.WriteString("        - \"$gostd\"\n")
		b.WriteString("    shouldNotDependsOn:\n")
		b.WriteString("      internal:\n")
		fmt.Fprintf(&b, "        - %q\n", fullPkg(modulePath, "application"))
		fmt.Fprintf(&b, "        - %q\n", fullPkg(modulePath, "infrastructure"))
		b.WriteString("\n")

		// Application layer: shouldNotDependsOn infrastructure
		b.WriteString("  # " + ctx.Name() + " application layer\n")
		fmt.Fprintf(&b, "  - package: %q\n", fullPkg(modulePath, "application"))
		b.WriteString("    shouldNotDependsOn:\n")
		b.WriteString("      internal:\n")
		fmt.Fprintf(&b, "        - %q\n", fullPkg(modulePath, "infrastructure"))
		b.WriteString("\n")
	}

	// Generate cross-context isolation rules
	contexts := bcMap.Contexts()
	for i, srcCtx := range contexts {
		for j, tgtCtx := range contexts {
			if i == j {
				continue
			}
			// Check if srcCtx is allowed to depend on tgtCtx
			if allowedDeps[srcCtx.Name()][tgtCtx.Name()] {
				continue // Relationship allows this dependency
			}
			// Generate isolation rule
			b.WriteString("  # " + srcCtx.Name() + " must not depend on " + tgtCtx.Name() + "\n")
			fmt.Fprintf(&b, "  - package: %q\n", fullPkg(srcCtx.ModulePath(), ""))
			b.WriteString("    shouldNotDependsOn:\n")
			b.WriteString("      internal:\n")
			fmt.Fprintf(&b, "        - %q\n", fullPkg(tgtCtx.ModulePath(), ""))
			b.WriteString("\n")
		}
	}

	return b.String(), nil
}

// RenderPytestarchTests renders pytestarch test file content.
func (s *FitnessTestSuite) RenderPytestarchTests() (string, error) {
	if len(s.contracts) == 0 {
		return "", fmt.Errorf("no contracts generated yet — call GenerateContracts() first: %w",
			domainerrors.ErrInvariantViolation)
	}

	var b strings.Builder
	b.WriteString("\"\"\"Auto-generated architecture fitness tests.\n\n")
	b.WriteString("Generated by alto from bounded context map.\n")
	b.WriteString("\"\"\"\n\n")
	b.WriteString("import pytest\n")
	b.WriteString("from pytestarch import get_evaluable_architecture, Rule\n\n\n")
	fmt.Fprintf(&b, "@pytest.fixture(scope=\"session\")\n")
	fmt.Fprintf(&b, "def evaluable():\n")
	fmt.Fprintf(&b, "    return get_evaluable_architecture(\".\", \"%s\")\n\n", s.rootPackage)

	for _, rule := range s.archRules {
		fnName := "test_" + SnakeCase(rule.name)
		b.WriteString("\n")
		fmt.Fprintf(&b, "def %s(evaluable):\n", fnName)
		fmt.Fprintf(&b, "    \"\"\"%s\"\"\"\n", rule.assertion)

		ctxModule := SnakeCase(rule.contextName)
		domainMod := s.rootPackage + "." + ctxModule + ".domain"
		infraMod := s.rootPackage + "." + ctxModule + ".infrastructure"

		b.WriteString("    rule = (\n")
		b.WriteString("        Rule()\n")
		b.WriteString("        .modules_that()\n")
		fmt.Fprintf(&b, "        .are_named(\"%s\")\n", infraMod)
		b.WriteString("        .should_not()\n")
		b.WriteString("        .be_imported_by_modules_that()\n")
		fmt.Fprintf(&b, "        .are_named(\"%s\")\n", domainMod)
		b.WriteString("    )\n")
		b.WriteString("    rule.assert_applies(evaluable)\n\n")
	}

	return b.String(), nil
}

// -- Private helpers --

func (s *FitnessTestSuite) validateModuleBoundaries() error {
	for _, contract := range s.contracts {
		expectedPrefix := s.rootPackage + "." + SnakeCase(contract.contextName)
		for _, module := range contract.modules {
			if !strings.HasPrefix(module, expectedPrefix) {
				return fmt.Errorf(
					"contract '%s' references module '%s' outside its bounded context '%s' (expected prefix: '%s'): %w",
					contract.name, module, contract.contextName, expectedPrefix,
					domainerrors.ErrInvariantViolation)
			}
		}
		for _, module := range contract.forbiddenModules {
			if !strings.HasPrefix(module, expectedPrefix) {
				return fmt.Errorf(
					"contract '%s' references forbidden module '%s' outside its bounded context '%s' (expected prefix: '%s'): %w",
					contract.name, module, contract.contextName, expectedPrefix,
					domainerrors.ErrInvariantViolation)
			}
		}
	}
	return nil
}

func buildContract(ct ContractType, contextName, modulePrefix string) Contract {
	switch ct {
	case ContractTypeLayers:
		return NewContract(
			contextName+" DDD layer contract",
			ContractTypeLayers,
			contextName,
			[]string{
				modulePrefix + ".infrastructure",
				modulePrefix + ".application",
				modulePrefix + ".domain",
			}, nil)
	case ContractTypeForbidden:
		return NewContract(
			contextName+" domain isolation",
			ContractTypeForbidden,
			contextName,
			[]string{modulePrefix + ".domain"},
			[]string{modulePrefix + ".infrastructure"})
	case ContractTypeIndependence:
		return NewContract(
			contextName+" independence",
			ContractTypeIndependence,
			contextName,
			[]string{modulePrefix}, nil)
	case ContractTypeAcyclicSiblings:
		return NewContract(
			contextName+" acyclic siblings",
			ContractTypeAcyclicSiblings,
			contextName,
			[]string{modulePrefix}, nil)
	default:
		return NewContract(
			contextName+" acyclic siblings",
			ContractTypeAcyclicSiblings,
			contextName,
			[]string{modulePrefix}, nil)
	}
}

func inferStrictness(contracts []Contract) ContractStrictness {
	types := make(map[ContractType]bool)
	for _, c := range contracts {
		types[c.contractType] = true
	}
	if len(types) >= 4 {
		return ContractStrictnessStrict
	}
	if types[ContractTypeLayers] {
		return ContractStrictnessModerate
	}
	return ContractStrictnessMinimal
}
